src/vs/code/node/cli.ts

```diff
@@ -80,46 +79,78 @@ export async function main(argv: string[]): TPromise<any> {
 			});
 		}
 
-		// If we are running with input from stdin, pipe that into a file and
-		// open this file via arguments. Ignore this when we are passed with
-		// paths to open.
-		let isReadingFromStdin: boolean;
+		let stdinWithoutTty: boolean;
 		try {
-			isReadingFromStdin = args._.length === 0 && !process.stdin.isTTY; // Via https://twitter.com/MylesBorins/status/782009479382626304
+			stdinWithoutTty = !process.stdin.isTTY; // Via https://twitter.com/MylesBorins/status/782009479382626304
 		} catch (error) {
 			// Windows workaround for https://github.com/nodejs/node/issues/11656
 		}
 
 		let stdinFilePath: string;
-		if (isReadingFromStdin) {
-			let stdinFileError: Error;
-			stdinFilePath = paths.join(os.tmpdir(), `code-stdin-${Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 3)}.txt`);
-			try {
-				const stdinFileStream = fs.createWriteStream(stdinFilePath);
-				resolveTerminalEncoding(verbose).done(encoding => {
-
-					// Pipe into tmp file using terminals encoding
-					const converterStream = iconv.decodeStream(encoding);
-					process.stdin.pipe(converterStream).pipe(stdinFileStream);
-				});
-
-				// Make sure to open tmp file
-				argv.push(stdinFilePath);
+		if (stdinWithoutTty) {
+
+			// Read from stdin: we require a single "-" argument to be passed in order to start reading from
+			// stdin. We do this because there is no reliable way to find out if data is piped to stdin. Just
+			// checking for stdin being connected to a TTY is not enough (https://github.com/Microsoft/vscode/issues/40351)
+			if (args._.length === 1 && args._[0] === '-') {
+
+				// remove the "-" argument when we read from stdin
+				args._ = [];
+				argv = argv.filter(a => a !== '-');
+
+				// prepare temp file to read stdin to
+				stdinFilePath = paths.join(os.tmpdir(), `code-stdin-${Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 3)}.txt`);
+				let stdinFileError: Error;
+				try {
+					const stdinFileStream = fs.createWriteStream(stdinFilePath);
+					resolveTerminalEncoding(verbose).done(encoding => {
+
+						// Pipe into tmp file using terminals encoding
+						const converterStream = iconv.decodeStream(encoding);
+						process.stdin.pipe(converterStream).pipe(stdinFileStream);
+					});
+
+					// Make sure to open tmp file
+					argv.push(stdinFilePath);
+
+					// Enable --wait to get all data and ignore adding this to history
+					argv.push('--wait');
+					argv.push('--skip-add-to-recently-opened');
+					args.wait = true;
+				} catch (error) {
+					stdinFileError = error;
```

Comment:
The only code which will throw is on line 105: `const stdinFileStream...`. Should that line be the only one wrapped in a `try catch` block?

Reply:
@joaomoreno fixed via https://github.com/Microsoft/vscode/commit/7dbdccbfb8242dc43d1838d8d97a6208079bb908